agents/magpie/magpie_agent.py

```diff
@@ -0,0 +1,388 @@
+    def grid(cls, xdim, ydim, ygap=0):
+        fp = cls()
+        xs, ys = range(xdim), range(ydim)
+        cmap = "red_cmap"
+        i = 0
+        for y in ys:
+            for x in xs:
+                name = f"channel_{i}"
+                _y = y
+                if ygap > 0:
+                    _y += 0.5 * (y // ygap)
+
+                fp.xs.append(x)
+                fp.ys.append(_y)
+                fp.rots.append(0)
+                fp.cnames.append(name)
+                fp.eqs.append(f"{name}")
+
+                fp.cmaps.append("red_cmap")
+                fp.templates.append("box")
+                i += 1
+
+        fp.num_dets = len(fp.xs)
+        fp.chan_mask = np.arange(fp.num_dets)
+        return fp
+
+    @classmethod
+    def from_tune_csv(cls, tune_csv, wafer_scale=50):
+        det_map = np.genfromtxt(
+            tune_csv, delimiter=',', skip_header=1, dtype=None
+        )
+
+        cmaps = {
+            90: "red_cmap",
+            150: "blue_cmap",
+        }
+        templates = {
+            90: 'csv_template_c0_p0',
+            150: 'csv_template_c1_p0',
+        }
+        fp = cls(ndets=len(det_map))
+        eq = '/ + 1 s {} 2'
+        map_inv = []  # Mapping from fp idx to channel number
+        for i, det in enumerate(det_map):
+            band, chan, freq, x, y, pol_angle = [
+                det[idx] for idx in [1, 2, 12, 17, 18, 19]
+            ]
+            try:
+                freq = int(freq)
+            except:
+                freq = 90
+            abschan = band * 512 + chan
+            map_inv.append(abschan)
+            fp.xs[i] = x * wafer_scale
+            fp.ys[i] = y * wafer_scale
+            fp.rots[i] = np.deg2rad(pol_angle)
+            fp.cnames[i] = f"chan_{abschan}"
+            fp.eqs[i] = eq.format(fp.cnames[i])
+            fp.cmaps[i] = cmaps[freq]
+            fp.templates[i] = templates[freq]
+        map_inv = np.array(map_inv)
+        # Mapping form channel no. to fp-idx
+        fp.chan_mask = np.full(np.max(map_inv)+1, -1)
+        for idx, chan in enumerate(map_inv):
+            fp.chan_mask[chan] = idx
+
+        return fp
+
+    @classmethod
+    def from_wafer_file(cls, wafer_file, wafer_scale=50):
+        import copy
+        import quaternionarray as qa
+        import toml
+
+        dets = toml.load(wafer_file)
+        fp = cls(ndets=len(dets))
+        band_idxs = {}
+        bands_seen = 0
+
+        xaxis = np.array([1., 0., 0.])
+        zaxis = np.array([0., 0., 1.])
+        cmaps = ["red_cmap", "blue_cmap"]
+
+        def det_coords(det):
+            quat =np.array(det['quat']).astype(np.float64)
+            rdir = qa.rotate(quat, zaxis).flatten()
+            ang = np.arctan2(rdir[1], rdir[0])
+            orient = qa.rotate(quat, xaxis).flatten()
+            polang = np.arctan2(orient[1], orient[0])
+            mag = np.arccos(rdir[2]) * 180 / np.pi
+            xpos = mag * np.cos(ang)
+            ypos = mag * np.sin(ang)
```

Comment:
Consider just xpos, ypos = rdir[:2] * 180/np.pi.  Using arccos to recover small angles (mag) has a lot of numerical error, since you're looking for the delta in cos(delta) ~ 1 - delta^2/2 + ... .